keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,833 @@
+}
+
+// FieldSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the request is not limited.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+// For the kube-apiserver:
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+type FieldSelectorAttributes struct {
+	// rawSelector is the serialization of a field selector that would be included in a query parameter.
+	// Webhook implementations are encouraged to ignore rawSelector.
+    // The kube-apiserver's SubjectAccessReview will parse the rawSelector. 
+	RawSelector string
+
+	// requirements is the parsed interpretation of a field selector.
+	// All requirements must be met for a resource instance to match the selector.
+	// Webhook implementations should handle requirements, but how to handle them is up to the webhook.
+	// Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
+	// are not understood.
+	Requirements []FieldSelectorRequirement
+}
+
+// LabelSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the request is not limited.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+// For the kube-apiserver:
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+type LabelSelectorAttributes struct {
+	// rawSelector is the serialization of a field selector that would be included in a query parameter.
+    // Webhook implementations are encouraged to ignore rawSelector.
+	// The kube-apiserver's SubjectAccessReview will parse the rawSelector. 
+	RawSelector string
```

Comment:
Talking about RawSelector, at the moment we only support Equals/NotEquals/Exists/NotExists in the url encoding, right? As we're looking into this, how do we feel about adding In/NotIn semantics like `labelSelector=foo=a|b|c`? I guess it's kind of impossible at this point, because we don't limit the content set on label values, so we cannot choose a character that can act as the separator.

Another thing I was thinking is `labelSelector=foo=a,foo=b,foo=c`, but that ends up being three different requirements, which ANDed together becomes an unsatisfiable requirements set (see the logic below).

I don't know, but it feels like if we start expanding the scope of these selectors, that's a practical issue that instantly arises for consumers.

Reply:
Currently every discrete requirement is AND'd when evaluated server-side.  I don't see a need to change that behavior.

Reply:
> Currently every discrete requirement is AND'd when evaluated server-side. I don't see a need to change that behavior.

Agreed. In fact if we do change it from AND we break the ability to expand operators in the future safely.

With AND semantics, a webhook can check for restrictions they understand, short-circuit when they find a limiting one they are satisfied by, ignore ones with operators they don't recognize and treat them as non-limiting, and fail in a safe direction in the presence of unrecognized operators.

> it feels like if we start expanding the scope of these selectors, that's a practical issue that instantly arises for consumers.

Given selectors are already part of our API, both in string and structured form, I don't think plumbing the current semantics into authz really expands their scope, especially if we give explicit instructions for how to treat unrecognized operators that makes it safe to accommodate operator expansion in the future.

Reply:
Sorry if my comment was unclear. Not saying we should move away from ANDed requirements, that should definitely be the case.

I was talking about the fact that within a single requirement, the structured YAML/JSON requirements are more expressible than their string encoding, as the `In` and `NotIn` operators do not convert to string form. So just pointing out that one hard part about using the current selector API in string form is that you cannot say "watch objects with labels/fields with key `foo` equals `a` or `b`.

So I was thinking if there are ways to start supporting that in the future, given that the API server is the canonical place for parsing these queries now, but I'm afraid if it isn't possible to start treating some characters as the "values separator" as a value could be almost anything.

Reply:
> I'm afraid if it isn't possible to start treating some characters as the "values separator" as a value could be almost anything.

This came up separately when discussion potential future changes to field selectors.  The problem is recognized, but the construction used here in []Requirements (recommended for webhook usage) gives us flexibility to change how field selectors are specified.